lib/src/list_extensions.dart

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Extension methods on common collection types.
+
+import "dart:collection";
+import "dart:math";
+
+import "algorithms.dart";
+import "algorithms.dart" as algorithms show shuffle; // Shadowed by declaration.
+import "equality.dart";
+
+extension ListExtensions<E> on List<E> {
+  /// Takes an action for each element.
+  ///
+  /// Calls [action] for each element along with the index in the
+  /// iteration order.
+  void forEachIndexed(void action(int index, E element)) {
+    for (var index = 0; index < length; index++) {
+      action(index, this[index]);
+    }
+  }
+
+  /// Maps each element and its index to a new value.
+  Iterable<R> mapIndexed<R>(R convert(int index, E element)) sync* {
```

Comment:
How much value are we buying by specializing this against `Iterable.mapIndexed`? Do we expect `this[index]` to be meaningfully different (faster?) than `for(var element in this)`?

Reply:
Yes, but maybe not by a lot.